### PR TITLE
Add AutoAugmentation to image classification training #3581

### LINKED ISSUE #3581: Add AutoAugmentation to image classification training
**Is your feature request related to a problem? Please describe.**

The image augmentation capabilities currently offered in this repository have room for enhancement, particularly given the various advanced techniques available in recent research.

Recent research papers have demonstrated superior techniques that could be beneficial:
[AutoAugment: Learning Augmentation Policies from Data](https://arxiv.org/abs/1805.09501)
[RandAugment: Practical automated data augmentation with a reduced search space](https://arxiv.org/abs/1909.13719)
[TrivialAugment: Tuning-free Yet State-of-the-Art Data Augmentation](https://arxiv.org/abs/2103.10158)
[AugMix: A Simple Data Processing Method to Improve Robustness and Uncertainty](https://arxiv.org/abs/1912.02781)

**Describe the use case**

Image augmentation aids in creating a richer training dataset via image transformations. While the current features serve the purpose, the integration of state-of-the-art AutoAugmentation techniques would elevate the capabilities of Ludwig.

**Describe the solution you'd like**

I recommend looking into the following libraries and solutions to integrate:
- [Torchvision](https://pytorch.org/vision/stable/transforms.html#auto-augmentation)
- [Google Brain AutoML](https://github.com/google/automl/blob/master/efficientdet/aug/autoaugment.py)

**Describe alternatives you've considered**

Checkout this list of [data augmentation libraries](https://github.com/AgaMiko/data-augmentation-review#computer-vision)

### PR DESCRIPTION
**Pull request by ethanreidel**
Added image auto augmentation functionality to diversify image datasets. Improves model generalization.
#3581 
pytest ludwig/tests/ludwig/augmentation/test_auto_augmentation.py

Closes #3581

### CODE REVIEW

**Comment by arnavgarg1** [comment]
Hey @ethanreidel! Thanks so much for raising this PR - on first looks, everything looks good, but will wait for others to review as well!

Were you able to benchmark performance before and after using trivial augmentation?

File: ludwig/features/image_feature.py
Comment (arnavgarg1) [inline ludwig/features/image_feature.py:103]: Hey @ethanreidel! Ludwig currently doesn't support Py 3.11 and beyond. I believe the `match` and `case` statements were only introduced in Py311, so this is probably why the tests are failing. Are you able to use conditionals instead?

Comment (ethanreidel) [reply ludwig/features/image_feature.py:103]: Hi @arnavgarg1. Yeah you're 100% correct. Will swap to conditionals. 👍

Comment (ethanreidel) [reply ludwig/features/image_feature.py:103]: Also @arnavgarg1, regarding the question you asked about benchmark performance. I have tested all three auto-augmentation methods (auto_augment, trivial_augment, and rand_augment) on the MNIST dataset example. The auto augment method tended to perform the best of the three, however, I have yet to test with other datasets.

Comment (arnavgarg1) [reply ludwig/features/image_feature.py]: @ethanreidel Nice! Good to know :) Great work on this and I'm excited for this to land!

File: ludwig/schema/metadata/configs/features.yaml
Comment (justinxzhao) [inline ludwig/schema/metadata/configs/features.yaml:412]: nit: Let's get rid of these spaces

File: tests/ludwig/augmentation/test_auto_augmentation.py
Comment (justinxzhao) [inline tests/ludwig/augmentation/test_auto_augmentation.py:26]: Can you add an assert to make sure that the augmented image is as you expect (at least for the shape)?

File: ludwig/features/image_feature.py
Comment (justinxzhao) [inline ludwig/features/image_feature.py:112]: Is the first dimension of `imgs` the batch dimension?

File: ludwig/features/image_feature.py
Comment (justinxzhao) [inline ludwig/features/image_feature.py:114]: Is it possible to vectorize the augmentation operation so that it operates on the `imgs` tensor all at once?

File: ludwig/features/image_feature.py
Comment (ethanreidel) [reply ludwig/features/image_feature.py:112]: Yes.

File: ludwig/features/image_feature.py
Comment (ethanreidel) [reply ludwig/features/image_feature.py]: Hi @justinxzhao. I just pushed the changes that you described in the PR review. Thanks for helping get this feature out!

File: ludwig/features/image_feature.py
Comment (arnavgarg1) [reply ludwig/features/image_feature.py:112]: nit: It would be helpful to add a type hint for the `imgs` argument being passed into the forward function

File: tests/ludwig/augmentation/test_auto_augmentation.py
Comment (arnavgarg1) [reply tests/ludwig/augmentation/test_auto_augmentation.py]: Agreed! We want some some sort of success criteria beyond the functions working

File: ludwig/features/image_feature.py
Comment (alexsherstinsky) [inline ludwig/features/image_feature.py:102]: @ethanreidel If not too much trouble, just from the coding style perspective, `return` statements are terminal, so we would not need `elif`.  Would it be possible for you to make that change?  Thanks a lot!

**Review by alexsherstinsky — CHANGES_REQUESTED** [review]
I made one small suggestion -- the rest looks great to me, and I will defer to @arnavgarg1 to make the final approval.  Thank you very much, @ethanreidel !

**Comment by arnavgarg1** [comment]
Really nice work! Going to merge as soon as tests finish running :)

**Pull request merged**
